rust/connlib/snownet/src/node.rs

```diff
@@ -902,7 +912,7 @@ where
             };
 
             if let Some(existing) = self.allocations.get_mut(server) {
-                existing.refresh(username, password, realm, now);
+                existing.update_credentials(username, password, realm, now);
```

Comment:
I didn't know there were passwords. How do those work?

Reply:
The full explanation is here: https://www.rfc-editor.org/rfc/rfc8489.html#page-22

The short story is, it is an HMAC-based scheme and we incorporate another secret value that is generated by the relay on startup and shared with the portal.

Reply:
There are also nonces to prevent replay attacks. All communication with the relays is UDP so there is no transport security. STUN has its own integrity mechanism using an HMAC that also incorporates the password. The relay only allows requests from known username-password combinations.

In our case, they are automatically derived from a secret so we don't need to actually usernames and passwords.